src/macros-by-example.adoc

```diff
@@ -0,0 +1,827 @@
+
+[{mrk}]
+----
+(*public* byte_array
+  [(*tagless uint8\...* bytes)]
+  [bytes])
+----
+
+Invocations of this macro are encoded (more or less) as a plain sequence of octets, because the
+macro definition constrains the argument shape such that nothing else is acceptable.  A text
+invocation is written using normal ints:
+
+[{mrk}]
+----
+(:byte_array 0 1 2 3 4 5 6 7 8) => [0, 1, 2, 3, 4, 5, 6, 7, 8]
+(:byte_array 9 -10 11)          => _**error**: -10 is not a valid uint8_
+(:byte_array 256)               => _**error**: 256 is not a valid uint8_
+----
+As above, Ion text doesn’t have syntax specifically denoting “8-bit unsigned integers”, so to
+keep text and binary capabilities aligned, the parser rejects invocations where an argument value
+exceeds the range of the binary-only type.
+
+The use of tagless types brings inherent tradeoffs which requires careful consideration, but in
+the right circumstances the density wins can be significant.
+
+
+=== Macro Shapes
+
+We can now introduce the final kind of input constraint, macro-shaped parameters.  To understand
+the motivation, consider modeling a scatter-plot as a list of points:
+
+----
+[{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}, …]
+----
+
+Lists like these exhibit a lot of repetition. Since we already have a `point` macro, we can
+eliminate a fair amount:
+
+----
+[(:point 3 17), (:point 395 23), (:point 15 48), (:point 2023 5), …]
+----
+
+This eliminates all the ``x``s and ``y``s, but leaves repeated macro invocations.  We can try to
+wrap this in another macro, but we find the type constraints insufficient, since the tightest we
+can go is `*struct*`, and things aren’t really any better:
+
+[{mrk}]
+----
+(*public* scatterplot [(*struct\...* points)]
+  [points])
+----
+----
+(:scatterplot (:point 3 17) (:point 395 23) (:point 15 48) (:point 2023 5) …)
+----
+
+What we’d like is to build eliminate the `point` calls and just write a stream of pairs, something
+like:
+
+----
+(:scatterplot (3 17) (395 23) (15 48) (2023 5) …)
+----
+
+We can achieve exactly that with a macro-shaped parameter, in which we use the `point` macro as a
+pseudo-type:
+
+[{mrk}]
+----
+(*public* scatterplot [(*point\...* points)]
+  [points])
+----
+----
+(:scatterplot (3 17) (395 23) (15 48) (2023 5) …)
+=>
+[{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}, …]
+----
+
+What’s happening here is that each argument S-expression like `(3 17)` is _implicitly an
+E-expression_ invoking the `point` macro.  The argument mirrors the shape of the inner macro,
+without repeating its name.  Further, expansion of the implied ``point``s happens automatically,
+so the overall behavior is just like the preceding struct-based variant and the `points`
+parameter produces a stream of structs.
+
+As the text notation here suggests, macro-shaped parameters are tagless, and the binary form also
+elides any opcodes mentioning `point` and just encodes its arguments with minimal delimiting.
```

Comment:
I don't think I understand that.  

Maybe this is an answer: You can't use an E-expression in an argument slot expecting a tagless type.  (The binary has no way to say that a macro is being invoked, since there's no tag.)

Reply:
Suppose we get rid of the `tagless` from in the `point` definition. Scatterplot still has a tagless macro-shaped parameter, but the `point` macro-shape it uses no longer has tagless parameters.
```
(public point
  [(int x), (int y)]
  {x: x, y: y})
(public scatterplot [(point\... points)]
  [points])
```
Is this allowed?

Reply:
Yes, that's allowed.  Each invocation of `point` will have a tagged, potentially annotated and/or null, integer.
```
(:scatterplot (1 null) (a::1 null.int) ((:macro_producing_one_int) 9))
```
... is valid